src/garage/tf/algos/rl2npo.py

```diff
@@ -0,0 +1,188 @@
+                         discount=discount,
+                         gae_lambda=gae_lambda,
+                         center_adv=center_adv,
+                         positive_adv=positive_adv,
+                         fixed_horizon=fixed_horizon,
+                         pg_loss=pg_loss,
+                         lr_clip_range=lr_clip_range,
+                         max_kl_step=max_kl_step,
+                         optimizer=optimizer,
+                         optimizer_args=optimizer_args,
+                         policy_ent_coeff=policy_ent_coeff,
+                         use_softplus_entropy=use_softplus_entropy,
+                         use_neg_logli_entropy=use_neg_logli_entropy,
+                         stop_entropy_gradient=stop_entropy_gradient,
+                         entropy_method=entropy_method,
+                         flatten_input=flatten_input)
+
+    def optimize_policy(self, itr, samples_data):
+        """Optimize policy.
+
+        Args:
+            itr (int): Iteration number.
+            samples_data (dict): Processed sample data.
+                See process_samples() for details.
+
+        """
+        if self._fit_baseline == 'before':
+            self._fit_baseline_with_data(samples_data)
+            samples_data['baselines'] = self._get_baseline_prediction(
+                samples_data)
+
+        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+        # Train policy network
+        logger.log('Computing loss before')
+        loss_before = self._optimizer.loss(policy_opt_input_values)
+        logger.log('Computing KL before')
+        policy_kl_before = self._f_policy_kl(*policy_opt_input_values)
+        logger.log('Optimizing')
+        self._optimizer.optimize(policy_opt_input_values)
+        logger.log('Computing KL after')
+        policy_kl = self._f_policy_kl(*policy_opt_input_values)
+        logger.log('Computing loss after')
+        loss_after = self._optimizer.loss(policy_opt_input_values)
+        tabular.record('{}/LossBefore'.format(self.policy.name), loss_before)
+        tabular.record('{}/LossAfter'.format(self.policy.name), loss_after)
+        tabular.record('{}/dLoss'.format(self.policy.name),
+                       loss_before - loss_after)
+        tabular.record('{}/KLBefore'.format(self.policy.name),
+                       policy_kl_before)
+        tabular.record('{}/KL'.format(self.policy.name), policy_kl)
+        pol_ent = self._f_policy_entropy(*policy_opt_input_values)
+        tabular.record('{}/Entropy'.format(self.policy.name), np.mean(pol_ent))
+
+        ev = np_tensor_utils.explained_variance_1d(samples_data['baselines'],
+                                                   samples_data['returns'],
+                                                   samples_data['valids'])
+        tabular.record('{}/ExplainedVariance'.format(self.baseline.name), ev)
+
+    def _get_baseline_prediction(self, samples_data):
```

Comment:
is there a reason this isn't a method in NPO? then you can do a lot less in the inheritance.

Reply:
NPO get baseline prediction in `BatchPolopt.process_samples()`, so it isn't really needed in NPO

Reply:
see my comment above about picking a consistent implementation style

If you're using inheritance from NPO, you shouldn't be re-implementing functionality from NPO under *different* names. Inheritance is not a tool for arbitrarily picking and choosing functions from one class into another. It's a tool for *specializing* types from more general to more specific.

This means if `Foo.speak()` exists and says "Hello", and you want to make `Bar.speak()` say "Hello, Anson", you shouldn't just define `Bar._speak()` and call that instead! Now `Bar` has both `Bar.speak()` and `Bar._speak()`!!! If I pass `bar = Bar()` to a function `make_it_speak()` which was defined for `Foo`, it will still call `Bar.speak()` and get "Hello", which is not what you wanted. You haven't created a more specific type, you've created a new type entirely!

Now if `Bar` can `walk()` and `Foo` can't `walk()`, then yes you can add `Bar.walk()`. `Bar` can still `speak()`, so `class Bar(Foo)` still applies.

In your situation, you either have to make Foo general-enough that you can effectively inherit from it (e.g. like adding options to NPO), or you can abandon inheritance and use composition instead. Since we've already discussed why composition is hard right now, I suggest you make NPO more flexible for your purposes.